src/tir/op/op.cc

```diff
@@ -112,12 +112,23 @@ void BinaryOpMatchTypes(PrimExpr& lhs, PrimExpr& rhs, Span span) {  // NOLINT(*)
     ICHECK(ltype.lanes() == rtype.lanes()) << "Cannot match type " << ltype << " vs " << rtype;
   }
   if (lhs.dtype() == rhs.dtype()) return;
-  // Only do very simple type coversion
+
+  // We keep casting pretty simple
+  // Two different floating point types will upconvert the lower bit floating point 
+  // to the same type as the higher bit version. E.g. fp16 + fp32 --> fp32 + fp32.
+  // Furthermore:
   // int->float, DataType::Int(32)->int(64)
   // require the types to be relatively consistent
   // This will the reduce amount code generated by operators
   // and also help user to find potential type conversion problems.
-  if (!lhs.dtype().is_float() &&
+  if (lhs.dtype().is_float() && rhs.dtype().is_float()) {
+    int max_num_bits = std::max(lhs.dtype().bits(), rhs.dtype().bits());
+    if (lhs.dtype().bits() != max_num_bits) {
+      lhs = cast(rhs.dtype(), lhs);
+    } else {
+      rhs = cast(lhs.dtype(), rhs);
+    }
```

Comment:
Looks like you will still add a cast to RHS even their bits are already the same?

```suggestion
    if (lhs.dtype().bits() < rhs.dtype().bits()) {
      lhs = cast(rhs.dtype(), lhs);
    } else if (lhs.dtype().bits() > rhs.dtype().bits()) {
      rhs = cast(lhs.dtype(), rhs);
    }
```

Reply:
If they have the same bits then it should hit the return statement above.

That being said your code is probably easier to read so I will replace it.

Done!